### ISSUE TITLE
Add protection against gha-creds-*.json leakage by default

### ISSUE THREAD

**Issue by tstromberg** [google-github-actions/auth#298]
### TL;DR

I recently was part of a security incident where the short-lived credentials produced by this action were accidentally leaked as part of a broken release process.  

While the fault was fully our own, as part of the postmortem I noticed the prescient README.md notice in your repository about having a .gitignore file in place to prevent accidentally committing credentials.

For the users who try this action without fully reading the documentation, I propose that there be further safeguards against credential leakage by default.

### Expected behavior

The credentials are stored in a location unlikely to be included by release automation. For instance, storing the credentials in memory, or a file location created by mktemp(). 

One could even go further and specify `/dev/shm` as a base directory to ensure that the credentials aren't written to disk, but I'm not sure there is a huge advantage there. Github Actions makes directory available as a tmpfs volume, as do all other Linux systems.

Alternatively, this action could verify the presence of a .gitignore entry before writing the credentials file out.

### Action YAML

```yaml
N/A
```

**Comment by sethvargo** [comment]
Hi @tstromberg 

Thank you for the feedback, and sorry to hear about your security incident. We've been down this path before (https://github.com/google-github-actions/auth/issues/275, https://github.com/google-github-actions/auth/issues/134, https://github.com/google-github-actions/auth/issues/123, https://github.com/google-github-actions/auth/issues/264), but there does not exist a location outside of the workspace that is shared with Docker-based containers in the GitHub Actions environment.

The original implementation used a tmpdir, but that created two challenges:

1.  On self-hosted runners, anything outside of the workspace _persists_, so future jobs on the same runner had access to creds. This was fixed by cleaning up the credential after each run, but there was still a rare chance that the GitHub Actions process would crash without cleanup, leaving the credential behind. This ultimately led to moving the credential back into the workspace.

2.  Docker-based actions and services do not have access to filepaths outside of the workspace. Customers wanted the action to work "out of the box" without needing to [mount custom filepaths into each container](https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container#mounting-volumes-in-a-container). This led to moving the credential back into the workspace.

> The credentials are stored in a location unlikely to be included by release automation. For instance, storing the credentials in memory, or a file location created by mktemp().
> 
> One could even go further and specify /dev/shm as a base directory to ensure that the credentials aren't written to disk, but I'm not sure there is a huge advantage there. Github Actions makes directory available as a tmpfs volume, as do all other Linux systems.

I agree, but as noted above, none of those mounts are shared by Docker container. Furthermore, GitHub **does not publish** any documentation or guarantees about volume mount points. [Back in 2019](https://github.com/orgs/community/discussions/25511#discussioncomment-3248176), the volumes were different than today, so we're extremely hesitant to use anything other than the workspace, which _must_ be mounted into the volume by default.

> Alternatively, this action could verify the presence of a .gitignore entry before writing the credentials file out.

I'm not in favor of this, since then the action would need to have ignore-file parsing logic, which is significant scope creep.

I understand the concern, but this behavior is extremely well-documented (twice in the README and once in the TROUBLESHOOTING guide) and is necessary to support the broad use cases out-of-the-box.

**Comment by sethvargo** [comment]
> The documentation states to add gha-creds-*.json to .gitignore so if we're speaking strictly about the .gitignore file then is there a significant amount of parsing logic necessary to check if there is a line that matches exactly gha-creds-*.json?

Greping for the literal string `gha-creds-*.json` is possible, but it's incomplete. All of the following would successfully ignore the creds, but would not be detected by a simple string search:

- `g*`
- `gh*`
- `gha*`
- `gha-*`
- ...
- `*.json`
- `*n`

**Comment by SteffenL** [comment]
My question was indeed about the literal string `gha-creds-*.json` and how that might be helpful without complicating things too much; however, I suppose implementing and enabling a requirement like that could break existing workflows in case they have ignored the file using patterns other than exactly `gha-creds-*.json`. An example of breakage could be that the action started throwing an error if that literal string wasn't found even when the file would've been ignored by another pattern.

Had I been the developer of this action then I would have preferred the action to not concern itself with the `.gitignore` file but I would also have liked to make it much harder for the file to leak by accident because that's a real concern for users of the action.

**Comment by SteffenL** [comment]
Right. I don't really have anything else to add at this time except perhaps the idea to not persist the credentials *anywhere* on the file system.

**Comment by fd** [comment]
This probably should be a configurable setting with a safe default.

**Comment by tstromberg** [comment]
I was part of another security incident that involved this unexpected behavior. Sure, you could argue that every user should fully read and implement the documentation for `google-github-actions`, but the truth is that the default behavior is unexpected.

I stand by my assertion that there should be a technical safeguard for this unexpected behavior: whether refusing to execute if the necessary `.gitignore` entry is missing or storing the creds elsewhere, or some other creative solution.

**Comment by tstromberg** [comment]
This just happened again in a new repository ...

**Comment by varunsh-coder** [comment]
I want to propose an alternate location to store the credentials file, such that it is available in both the host and docker actions, but not in the working directory. Would love to get feedback on this proposal. 

1. Store the creds file in the `/github/home` directory on the host. 
2. Also store the creds file in the `_github_home` directory created on the host within its temp directory. This location gets mounted as `/github/home` in the docker container for docker actions. [Reference to the code](https://github.com/actions/runner/blob/36c66c808363839ea7069bef87fd4572c8bb6c54/src/Runner.Worker/Handlers/ContainerActionHandler.cs#L194) in the runner repo. 
3. Set the environment variables that point to the creds files to `/github/home/credsfile`. This file should now be present on both the host and the docker container.

**Comment by sethvargo** [comment]
Unless it's a public, documented API in the GitHub support documentation, we cannot use it. We've been repeatedly burned by depending on behavior identified in the OSS runner code that changes without notice, and the feedback from GitHub is that it's not part of a stable API and cannot be depended on.
